File: internal/domain/progress.go
package domain

import "time"

// HistoryEntry is one data point in an exercise's weight history.
// It represents the best (heaviest) eligible set logged in a single
// completed session. "Eligible" means actual_reps >= target_reps for
// weight_reps exercises; all sets count for other tracking types.
type HistoryEntry struct {
	SessionID   int64 // internal ID used for cursor pagination
	SessionUUID string
	CompletedAt time.Time
	Weight      float64
}

// PersonalRecord holds the heaviest eligible set ever logged for an exercise.
type PersonalRecord struct {
	Weight      float64
	ActualReps  *int
	SessionUUID string
	CompletedAt time.Time
}

// PersonalRecordListEntry holds a PR for one exercise, used in the batch PR list.
type PersonalRecordListEntry struct {
	ExerciseUUID string
	ExerciseName string
	Weight       float64
	ActualReps   *int
	CompletedAt  time.Time
}

// ExerciseChartPoint is one data point for an exercise progress chart.
// Represents the best (heaviest eligible) set per completed session.
type ExerciseChartPoint struct {
	Date   string  // ISO date (YYYY-MM-DD)
	Weight float64
	Reps   int
	Volume float64 // weight * reps
}

// RecentSession is a completed or skipped session with context for display.
type RecentSession struct {
	UUID        string
	CycleUUID   string
	Status      SessionStatus
	CompletedAt *time.Time
	WorkoutName string
	ProgramName string
}

// ProgressSummary holds aggregate stats for all sessions across all cycles.
type ProgressSummary struct {
	TotalSessions int64
	WeeksTrained  int64
	CurrentStreak int
}

File: app/domain/progress.ts
// Progress read models — mirrors internal/domain/progress.go.
// These are pure data types with no business logic.

export interface HistoryEntry {
  session_id: number;
  session_uuid: string;
  completed_at: string;
  weight: number;
}

export interface PersonalRecord {
  weight: number;
  actual_reps: number | null;
  session_uuid: string;
  completed_at: string;
}

export interface PersonalRecordListEntry {
  exercise_uuid: string;
  exercise_name: string;
  weight: number;
  actual_reps: number | null;
  completed_at: string;
}

export interface ExerciseChartPoint {
  date: string;
  weight: number;
  reps: number;
  volume: number;
}

export interface RecentSession {
  uuid: string;
  cycle_uuid: string;
  status: string;
  completed_at: string | null;
  workout_name: string;
  program_name: string;
}

export interface ProgressSummary {
  total_sessions: number;
  weeks_trained: number;
  current_streak: number;
}
